src/globus_sdk/experimental/globus_app/globus_app.py

```diff
@@ -0,0 +1,429 @@
+
+        self.client_id = client_id
+        self.client_secret = client_secret
+        super().__init__(app_name, scope_requirements=scope_requirements, config=config)
+
+        # either get config's LoginFlowManager
+        # or make a default CommandLineLoginFlowManager
+        if self.config.login_flow_manager:
+            self._login_flow_manager = self.config.login_flow_manager
+        else:
+            self._login_flow_manager = CommandLineLoginFlowManager(
+                self._login_client,
+                refresh_tokens=self.config.refresh_tokens,
+            )
+
+    def _initialize_login_client(self) -> None:
+        if self.client_secret:
+            self._login_client = ConfidentialAppAuthClient(
+                app_name=self.app_name,
+                client_id=self.client_id,
+                client_secret=self.client_secret,
+            )
+        else:
+            self._login_client = NativeAppAuthClient(
+                app_name=self.app_name,
+                client_id=self.client_id,
+            )
+
+    def _initialize_authorizer_factory(self) -> None:
+        if self.config.refresh_tokens:
+            self._authorizer_factory = RefreshTokenAuthorizerFactory(
+                token_storage=self._validating_token_storage,
+                auth_login_client=self._login_client,
+            )
+        else:
+            self._authorizer_factory = AccessTokenAuthorizerFactory(
+                token_storage=self._validating_token_storage,
+            )
+
+    def run_login_flow(
+        self, auth_params: GlobusAuthorizationParameters | None = None
+    ) -> None:
+        """
+        Run an authorization flow to get new tokens which are stored and available
+        for the next authorizer gotten by get_authorizer.
+
+        As a UserApp this always involves an interactive login flow with the user
+        driven by the app's LoginFlowManager.
+
+        :param auth_params: A GlobusAuthorizationParameters to control how the user
+            will authenticate. If not passed
+        """
+        auth_params = self._auth_params_with_required_scopes(auth_params)
+        token_response = self._login_flow_manager.run_login_flow(auth_params)
+        self._authorizer_factory.store_token_response_and_clear_cache(token_response)
+
+
+class ClientApp(GlobusApp):
+    """
+    A ``GlobusApp`` using client credentials - useful for service accounts and
+    automation use cases.
+
+    .. tab-set::
+
+        .. tab-item:: Future Example Usage
+
+            .. code-block:: python
+
+                app = UserApp("myapp", CLIENT_ID, CLIENT_SECRET)
+                client = TransferClient(app=app)
+                res = client.operation_ls(COLLECTION_ID)
+
+    """
+
+    _login_client: ConfidentialAppAuthClient
+    _authorizer_factory: ClientCredentialsAuthorizerFactory  # type:ignore
+
+    def __init__(
+        self,
+        app_name: str,
+        client_id: UUIDLike,
+        client_secret: str,
```

Comment:
Given that these could become optional in the future (with environment variables) can we move them to the kwargs & validate their presence in runtime code instead?

Reply:
For some reason I thought client environment variables would only be for a `UserApp`, but not sure where that idea came from